File: Soal4/210015_UTSPBO_Soal4.cpp
/*
* Nama :Amir Salim
* NPM : 140810210015
* Kelas : A
* Tanggal : 19 Oktober 2022
* Nama Program : 210015_UTSPBO_Soal4.cpp
* Deskripsi :  UTS SOAL 4
*/

#include<iostream>
#include<string>


class Waktu{
    private:
        int jam,menit,detik;
    
    public:
        Waktu(int jam,int menit,int detik){
            this->jam = jam;
            this->menit = menit;
            this->detik = detik;
        }
        Waktu(){
            this->jam = 0;
            this->menit =0;
            this->detik=0;
        }
        void setJam(int jam){
            this->jam = jam;
        }
        void setMenit(int menit){
            this->menit = menit;
        }
        void setDetik(int detik){
            this->detik = detik;
        }
        void inputWaktu(){
            std::cout<<"Masukkan jam : ";
            std::cin>>this->jam;

            std::cout<<"Masukkan menit : ";
            std::cin>>this->menit;

            std::cout<<"Masukkan detik : ";
            std::cin>>this->detik;
        }
        int getJam(){
            return this->jam;
        }
        int getMenit(){
            return this->menit;
        }
        int getDetik(){
            return this->detik;
        }
        std::string getWaktu(){
            std::string nolJam ="";
            std::string nolMenit="";
            std::string nolDetik="";

            if(this->jam<10){
                nolJam="0";
            }
            if(this->menit<10){
                nolMenit="0";
            }
            if(this->detik<10){
                nolDetik="0";
            }
            
            
            return nolJam + std::to_string(this->jam) + ":" + nolMenit+ std::to_string(this->menit) + ":" +nolDetik+ std::to_string(this->detik);
        }
        int convertToSecond(){
            int hasil = this->detik + this->menit*60 + this->jam*3600;
            return hasil;
        }
        void secondToClock(int second){
            this->menit=second/60;
            this->detik=second%60;
            this->jam=this->menit/60;
            this->menit=this->menit%60;
        }
        Waktu cariDurasi(Waktu akhir){
             Waktu temp;

            int detikAwal = this->convertToSecond();
            int detikAkhir = akhir.convertToSecond();
            if(detikAkhir<detikAwal){
                detikAkhir+=86400;
            }
            int detikHasil = detikAkhir - detikAwal;

            

            temp.secondToClock(detikHasil);
            return temp;
        }
};

class Parkir{
    private:
        std::string no;
        int jenis;
        Waktu datang;
        Waktu pulang;
    
    public:
        Parkir(){
            this->no = " ";
            this->jenis=0;
        }

        //Input
        void setNoKendaraan(std::string no){
            this->no = no;
        }

        void setJenis(int jenis){
            this->jenis = jenis;
        }

        void setWaktuDatang(Waktu datang){
            this->datang = datang;

        }

        void setWaktuPulang(Waktu pulang){
            this->pulang=pulang;
        }

        void inputKendaraan(){
            std::cout<<"\n--- INPUT KENDARAAN ---\n";
            std::cout<<"No Kendaraan : ";
            std::cin.ignore();
            std::getline(std::cin,this->no);

            std::cout<<"Jenis Kendaraan : ";
            std::cin>>jenis;

            std::cout<<"\n-- Jam Masuk Kendaraan --\n";
            this->datang.inputWaktu();

            std::cout<<"\n-- Jam Keluar Kendaraan --\n";
            this->pulang.inputWaktu();
        }

        //Output  
        std::string getNoKendaraan(){
            return this->no;
        }

        int getJenis(){
            return this->jenis;
        }

        Waktu getWaktudatang(){
            return this->datang;
        }

        Waktu getWaktuPulang(){
            return this->pulang;
        }

        //Proses 
        std::string getNamaJenis(){
             std::string hasil = " ";

            if(this->jenis == 1){
            hasil = "Motor";
            }
            else if(this->jenis==2){
            hasil = "Mobil";
            }

            return hasil;
        }

        Waktu getLamaParkir(){
            return this->datang.cariDurasi(this->pulang);
        }

        int getLamaJam(){
        int hasil = 0;
            if(this->getLamaParkir().getMenit()>=10 || this->getLamaParkir().getJam()>=1){
                hasil = this->getLamaParkir().getJam();

                if( this->getLamaParkir().getMenit()>0 || this->getLamaParkir().getDetik()>0 ){
                    hasil +=1;
                }
           
            }
            return hasil;
        }

        int getBiayaParkir(){
            int hasil = 0;
            if(this->jenis==1){
                hasil = this->getLamaJam() * 2000;
            }

            else if(this->jenis==2){
                hasil = this->getLamaJam() * 3000;
            }

        return hasil;
    }    

};

class LarikParkir{
    private:
        int ukuran;
        Parkir p[100];
    
    public:
        LarikParkir(int ukuran){
            this->ukuran = ukuran;
        }

        void inputLarikParkir(){
                for(int i=0;i<this->ukuran;i++){
                    system("cls");
                    std::cout<<"Kendaraan ke - " <<(i+1);
                    this->p[i].inputKendaraan();
                    system("cls");
            }
        }

        void cetakTabelParkir(){
            std::cout<<"\t\t\t\tRekapitulasi Biaya parkir PT Parkir Jaya\n";

            if(this->p[0].getNoKendaraan()=="s"){
                std::cout<<"Larik Kosong !\n";

            }

            else{
                int no = 1;
                std::cout<<"================================================================================================================================\n";
                std::cout<<"No\tNo Kendaraan\t\tJenis\tJam Masuk\tJam Pulang\tLama Parkir\tLama jam\tBiaya\n";
                std::cout<<"================================================================================================================================\n";

                for(int i=0;i<this->ukuran;i++){
                    if(this->p[i].getNoKendaraan() == " "){
                        break;
                    }

                    else{
                        std::cout<<no<<"\t"<<
                        this->p[i].getNoKendaraan() << "\t\t" << 
                        this->p[i].getNamaJenis()<< "\t" << 
                        this->p[i].getWaktudatang().getWaktu() << "\t" << 
                        this->p[i].getWaktuPulang().getWaktu() << "\t" <<   
                        this->p[i].getLamaParkir().getWaktu() << "\t " << 
                        this->p[i].getLamaJam() << "\t\t" << 
                        this->p[i].getBiayaParkir() << "\t  \n";  
                    }
                    no++;
                }
                std::cout<<"================================================================================================================================\n";
            }
        }

        float totalBiaya(){
            float hasil = 0;
            for(int i=0;i<this->ukuran;i++){
                hasil+=this->p[i].getBiayaParkir();
            }

            return hasil;
        }

};

int main()
{
    int ukuran;
    std::cout<<"Masukkan ukuran matriks = ";
    std::cin>>ukuran;

    LarikParkir lp(ukuran);

    system("cls");

    lp.inputLarikParkir();
    system("cls");
    lp.cetakTabelParkir();

    std::cout<<"Total Biaya adalah = "<<lp.totalBiaya();

    return 0;
}

File: Soal4/210015_UTSPBO_Soal4.py
# Nama :Amir Salim
# NPM : 140810210015
# Kelas : A
# Tanggal : 19 Oktober 2022
# Nama Program : 210015_UTSPBO_Soal4.py
# Deskripsi :  UTS SOAL 4

from os import system

class Waktu:

    #Attribute
    __jam=0
    __menit=0
    __detik=0

    #Constructor
    def __init__(self, *args):
          if (len(args) == 3):
            self.__jam = int(args[0])
            self.__menit = int(args[1])
            self.__detik = int(args[2])
          
          elif(len(args)==0):
            self.__jam = int(0)
            self.__menit = int(0)
            self.__detik = int(0)

          else:
            print("False number of argument in constructor") 

    #Input Method
    def setJam(self,jam):
        self.__jam = int(jam)
    
    def setMenit(self,menit):
        self.__menit = int(menit)
    
    def setDetik(self,detik):
        self.__detik = int(detik)
    
    def inputWaktu(self):
       
        self.__jam = int(input("Masukkan jam : "))
        self.__menit = int(input("Masukkan menit : "))
        self.__detik = int(input("Masukkan detik : "))
    
    #Output Method
    def getJam(self):
        return self.__jam
    
    def getMenit(self):
        return self.__menit
    
    def getDetik(self):
        return self.__detik
    
    def getWaktu(self):
        nolJam =""
        nolMenit=""
        nolDetik=""

        if(self.__jam<10):
            nolJam="0"
            
        if(self.__menit<10):
            nolMenit="0"
            
        if(self.__detik<10):
            nolDetik="0"

        return nolJam + str(self.__jam) + ":" + str(nolMenit)+ str(self.__menit) + ":" +nolDetik+ str(self.__detik)
    
    #Proses

    def convertToSecond(self):
        hasil = self.__detik + (int(60) * self.__menit) + (int(3600) * self.__jam)
        return hasil 
    
    def secondToClock(self,second:int):
        self.__menit = int(second/60)
        self.__detik = int(second%60)
        self.__jam = int(self.__menit/60)
        self.__menit = int(self.__menit%60)
    
    def cariDurasi(self,akhir):
        temp = Waktu()

        detikAwal = self.convertToSecond()
        detikAkhir = akhir.convertToSecond()

        if(detikAkhir<detikAwal):
            detikAkhir+=86400
        
        detikHasil = detikAkhir - detikAwal
        temp.secondToClock(detikHasil)

        return temp

#-----------------------------------------------------------------------------------------------------------------------------------------

class Parkir:
    __no = " "
    __jenis=int(0)
    __datang = Waktu()
    __pulang = Waktu()

    #Constructor
    def __init__(self):
        self.__no = " "
        self.__jenis = 0
        self.__datang = Waktu(0,0,0)
        self.__pulang = Waktu(0,0,0)


    #Input
    def setNoKendaraan(self,no):
        self.__no = no
    
    def setJenis(self,jenis):
        self.__jenis=jenis
    
    def setWaktuDatang(self,datang):
        self.__datang = datang
    
    def setWaktuPulang(self,pulang):
        self.__pulang = pulang
    
    def inputKendaraan(self):
        print("\n--- INPUT KENDARAAN---")
        self.__no = input("No Kendaraan : ")
        self.__jenis = int(input("Jenis Kendaraan : "))

        print("\n--- Jam Masuk Kendaraan ---")
        self.__datang.inputWaktu()

        print("\n--- Jam Keluar Kendaraan ---")
        self.__pulang.inputWaktu()
    
    #Output
    def getNoKendaraan(self):
        return self.__no

    def getJenis(self):
        return self.__jenis
    
    def getWaktuDatang(self):
        return self.__datang
    
    def getWaktuPulang(self):
        return self.__pulang
    
    def getNamaJenis(self):
        hasil = " "

        if(self.__jenis == 1):
            hasil = "Motor"
        elif(self.__jenis==2):
            hasil="Mobil"
        
        return hasil
    
    def getLamaParkir(self):
        return self.__datang.cariDurasi(self.__pulang)
    
    def getLamaJam(self):
        hasil = int(0)
        if(self.getLamaParkir().getMenit()>=10 or self.getLamaParkir().getJam()>=1):
            hasil = self.getLamaParkir().getJam()

            if(self.getLamaParkir().getMenit()>0 or self.getLamaParkir().getDetik()>0):
                hasil = hasil + 1
        
        return hasil
    
    def getBiayaParkir(self):
        hasil = int(0)
        if(self.__jenis==1):
            hasil = self.getLamaJam()*2000
        
        elif(self.__jenis==2):
            hasil = self.getLamaJam()*3000
        
        return hasil

#-----------------------------------------------------------------------------------------------------------------------------------------
class LarikParkir:
    __ukuran=int(0)
    __p = []

    #Constructor
    def __init__(self,ukuran):
        self.__ukuran = ukuran
    
    #Input
    def inputLarikParkir(self):
        i = int(0)
        while(i<self.__ukuran):
            system("cls")
            print("Kendaraan ke - ",(i+1))
            obj = Parkir()
            obj.inputKendaraan()
            self.__p.append(obj)
            system("cls")
            i=i+1
    
    #Output
    def cetakTabelParkir(self):
        print("\t\t\t\tRekapitulasi Biaya parkir PT Parkir Jaya\n")
        if(self.__p[0].getNoKendaraan()==" "):
            print("Larik Kosong ! ")
        
        else:
            no = int(1)
            i = int(0)
            print("================================================================================================================================")
            print("No\tNo Kendaraan\t\tJenis\tJam Masuk\tJam Pulang\tLama Parkir\tLama jam\tBiaya");
            print("================================================================================================================================")
            while(i<self.__ukuran):
                if(self.__p[i].getNoKendaraan() == " "):
                    break

                else:
                    print(
                    no,"\t",
                    self.__p[i].getNoKendaraan(),"\t\t", 
                    self.__p[i].getNamaJenis(),"\t", 
                    self.__p[i].getWaktuDatang().getWaktu(),"\t", 
                    self.__p[i].getWaktuPulang().getWaktu(),"\t",   
                    self.__p[i].getLamaParkir().getWaktu(),"\t ", 
                    self.__p[i].getLamaJam(),"\t\t", 
                    self.__p[i].getBiayaParkir(),"\t  "  
                         )
                no = no+1;
                i = i+1
            print("================================================================================================================================")
    
    #Proses
    def totalBiaya(self):
        hasil = float(0)
        i = int(0)
        while(i<self.__ukuran):
            hasil = hasil + self.__p[i].getBiayaParkir()
            i = i + 1
        return hasil


#-----------------------------------------------------------------------------------------------------------------------------------------

uk = int(input("Masukkan ukuran matriks : "))

system("cls")

lp = LarikParkir(uk)

lp.inputLarikParkir()
system("cls")
lp.cetakTabelParkir()

print("Total Biaya Parkir adalah = ",lp.totalBiaya())